README.md
# GENEMABR: Gene-set module annotation or gene-set enrichment within a regression based framework

# Introduction

GENEMABR is  an R package for gene module annotation or gene set enrichment analysis via a regression based method.
We formulate the gene set enrichment problem within a regression framework. Thus, the problem of gene-set enrichment is transformed into a feature selection problem, where the aim is to select the gene sets that best predict the membership of genes in a given gene set/module.
Here we propose to apply regularised regression methods, such as lasso (l1 regularization), ridge (l2 regularization), or elastic net (hybrid of l1 and l2 regularization controlled by the hyperparameter alpha), in order to adjust the treatment of similar or redundant gene sets.
For more details about this method. Please refer to out paper:(?)

# Installation Instructions

To install the most recent release of GENEMABR (R >= 3.6) via devtools:
 ```R
 library(devtools)
 install_github("TaoDFang/GENEMABR")
 ```

 To install GENEMABR via bioConductor:
 Not for now


When you load library by library(GENEMABR)
if you see error:
Error in fetch(key) :
  lazy-load database '/Library/Frameworks/R.framework/Versions/3.6/Resources/library/GENEMABR/help/GENEMABR.rdb' is corrupt

You can restart R session to solve the problem

# Examples and vignettes

Check pdf format file in  vignettes folder


NEWS
Changes in version 0.99.0 (2019-05-11)
+ Submitted to Bioconductor


tests/testthat.R
library(testthat)
library(GENEMABR)

test_check("GENEMABR")


tests/testthat/test-supplementary_functions.R
library(GENEMABR)
library(igraph)
library(Matrix)


test_that("Test find_root_ids", {
    a=list()
    a[[1]]="R-HSA-168256"
    names(a)="R-HSA-1810476"

    b=list()
    b[[1]]="GO:0065007"
    names(b)="GO:0032688"

    expect_equal(find_root_ids("R-HSA-1810476"), a)
    expect_equal(find_root_ids("GO:0032688"), b)
})


test_that("Test from_id2name", {
    a=list()
    a[[1]]="R-HSA-168256"
    names(a)="R-HSA-1810476"

    b=list()
    b[[1]]="GO:0009987#GO:0065007"
    names(b)="GO:0007249"


    test_a=list()
    test_a[[1]]="Immune System"
    names(test_a)="R-HSA-168256"

    test_b=list()
    test_b[[1]]=c("cellular process", "biological regulation")
    names(test_b)="GO:0009987#GO:0065007"

    expect_equal(from_id2name(a), test_a)
    expect_equal(from_id2name(b), test_b)
})


test_that("Test get_steps", {
    a=list(3,4)
    names(a)=c("GO:0005834","R-HSA-111469")
    expect_equal(get_steps(c("GO:0005834","R-HSA-111469")), a)

})


tests/testthat/test-regression_selected_pathways_function.R
library(GENEMABR)
library(glmnet)
library(Matrix)

test_that("Test regression_selected_pathways", {

    # selected_pathways_names=list("RIP-mediated NFkB activation via ZBP1","IkBA variant leads to EDA-ID",
    #                              "negative regulation of interferon-beta production","TRAF6 mediated NF-kB activation",
    #                              "I-kappaB kinase/NF-kappaB signaling","ZBP1(DAI) mediated induction of type I IFNs")
    # names(selected_pathways_names)=c("R-HSA-1810476","R-HSA-5603029","GO:0032688","R-HSA-933542","GO:0007249","R-HSA-1606322")
    #
    # selected_pathways_coef=c(0.12593491,0.07105989,0.01993546,0.01910265,0.00513055, 0.00219621 )
    # names(selected_pathways_coef)=c("R-HSA-1810476","R-HSA-5603029","GO:0032688","R-HSA-933542","GO:0007249","R-HSA-1606322")
    #
    # selected_pathways_fisher_pvalue=c(3.511133e-10,4.005332e-08,6.338554e-05,1.929685e-09,4.985802e-11,7.596824e-10)
    # names(selected_pathways_fisher_pvalue)=c("R-HSA-1810476","R-HSA-5603029","GO:0032688","R-HSA-933542","GO:0007249","R-HSA-1606322")
    #
    # selected_pathways_num_genes=c( 11,7,11,16 ,63,13)
    # names(selected_pathways_num_genes)=c("R-HSA-1810476","R-HSA-5603029","GO:0032688","R-HSA-933542","GO:0007249","R-HSA-1606322")
    #
    # a=list(selected_pathways_names=selected_pathways_names,selected_pathways_coef=selected_pathways_coef,
    #        selected_pathways_fisher_pvalue=selected_pathways_fisher_pvalue,selected_pathways_num_genes=selected_pathways_num_genes)
    #
    # expect_equal(regression_selected_pathways(gene_input = c("TRPC4AP","CDC37","TNIP1","IKBKB","NKIRAS2","NFKBIA","TIMM50","RELB",
    #                                                          "TNFAIP3","NFKBIB","HSPA1A","NFKBIE","SPAG9",
    #                                                          "NFKB2","ERLIN1","REL","TNIP2","TUBB6","MAP3K8"),alpha = 0.5), a)

    expect_equal(regression_selected_pathways(gene_input = c("TRPC4AP","CDC37","TNIP1"),alpha = 1), NULL)
})


R/data.R
#' Homo sapiens GO ontology tree
#'@description A rds R object contains GO ontology relationships (tree structure)
#'@format Directed igraph format
#'@source http://geneontology.org/docs/download-ontology/
#'@details  The igraph format tree was constructed by using data from http://geneontology.org/docs/download-ontology/ (May 2108)
#'It has three root notes representing Molecular Function,Cellular Component and Biological Process (http://geneontology.org/docs/ontology-documentation/)
"human_go_ontology"


#' human_go_roots
#'@description A rds R object contains GO ontology tree root nodes
#'@format A vector of GO ontology root notes (ID)
#'@source http://geneontology.org/docs/download-ontology/
#'@details  You can view tree stuctor of GO ontology at  https://www.ebi.ac.uk/QuickGO/
#'Thre are three roots notes in GO ontology tree: GO:0008150 (biological_process), GO:0003674(molecular_function), GO:0005575(cellular_component)
"human_go_roots"

#' human_go_sub_roots
#'@description A rds R object contains GO ontology tree sub-root nodes (The children of root nodes).
#'@format A list of three elements contains GO ontology sub-root notes (ID)/the children of three root notes
#'@source http://geneontology.org/docs/download-ontology/
#'@details  You can view tree stuctor of GO ontology at  https://www.ebi.ac.uk/QuickGO/
"human_go_sub_roots"


#' Homo sapiens REACTOME ontology tree
#'@description A rds R object contains Reactome ontology relationships (tree structure)
#'@format Directed igraph format
#'@source https://reactome.org/download-data
#'@details  The igraph format tree was constructed by using  data from https://reactome.org/download-data (May 2108)
#'It has several root nodes representing REACTOME pathway categories (https://reactome.org/PathwayBrowser/)
"human_reactome_ontology"


#' human_reactome_roots
#'@description A rds R object contains REACTOME tree root nodes
#'@format A vector of REACTOME root notes (ID)
#'@source https://reactome.org/download-data
#'@details  You can view tree stuctor of REACTOME at https://reactome.org/PathwayBrowser/
"human_reactome_roots"


#' Homo sapiens GO ontology and REACTOME ontology gene-pathway realtionship
#'@description A rds R object contains GO ontology and REACTOME ontology gene-pathway realtionship
#'@format Formal class 'dgCMatrix' [package "Matrix"]
#'@source http://geneontology.org/docs/download-ontology/, https://reactome.org/download-data
#'A binary matrix whose columns are the pathways/gene sets from GO ontology and REATOME database and
#'whose rows are all the genes(represented by gene HUGO gene symbols) from  GO ontology and REATOME database.
#'For gene i and pathway j, the value of matrix(i,j) is 1 is gene i belonging to pathway j otherwise 0
"gene_pathway_matrix"



R/regression_selected_pathways_funcs.R
#' regression_selected_pathways
#'
#' This function allows you to extracte enriched pathways for gene module/list via regressioin (elastic net) based method
#' @param gene_input A vecor of genes to be annotated. It should have same ID types(Ensembl ID, HUGO gene symbol) as the genes in \emph{gene_pathway_matrix}.
#' @param gene_pathway_matrix A binary background matrix whose columns are the pathways/gene sets and
#'whose rows are all the genes from pathways/gene sets . It could be in sparse matrix format ((inherit from class "sparseMatrix" as in package Matrix) to save memory.
#'For gene i and pathway j, the value of matrix(i,j) is 1 is gene i belonging to pathway j otherwise 0.
#'Users could leave it as default value("default") so it will use pre-collected gene_pathway_matrix from GO Ontology and REACTOME databaase.
#'Otherwise, they could use their own customized gene_pathway_matrix
#' @param lambda We use glmnet function to do regression. \emph{lambda} is an argument in \strong{glmnet}. See \strong{glmnet} function for more details
#' Here we use default value 0.007956622 after preliminary study. It can be overridden by giving \emph{nlambda} and \emph{lambda.min.ratio arguments}.
#' @param alpha The elasticnet mixing parameter, with \eqn{0~\leq~\alpha~\leq~1}. The penalty is defined as
#'\eqn{(1-\alpha)/2||\beta||_2^2+\alpha||\beta||_1}.
#'alpha=1 is the lasso penalty, and alpha=0 the ridge penalty. Default value: 0.5.
#' @param ... Other paramaters for glmnet function.
#' @return  A list of four elements:
#' \itemize{
#'   \item selected_pathways_names - Pathways names for selected pathways
#'   \item selected_pathways_coef - Regression coefficients value for selected pathways
#'   \item selected_pathways_fisher_pvalue - Fisher exact pvalue for selected pathways
#'   \item selected_pathways_num_genes - The number of genes for selected pathways in background
#' }
#' @importFrom glmnet glmnet
#' @export
#' @examples
#' rspResults <- regression_selected_pathways(gene_input=c("TRPC4AP","CDC37",
#'   "TNIP1","IKBKB","NKIRAS2", "NFKBIA","TIMM50","RELB","TNFAIP3","NFKBIB",
#'   "HSPA1A","NFKBIE","SPAG9","NFKB2","ERLIN1","REL","TNIP2",
#'   "TUBB6","MAP3K8"),
#'  gene_pathway_matrix="default",lambda=0.007956622,alpha=0.5)
regression_selected_pathways=function(gene_input,gene_pathway_matrix="default",lambda=0.007956622,alpha=0.5,...){
  addi_args=list(...)
  if(gene_pathway_matrix=="default"){
     gene_pathway_matrix <- mydata("gene_pathway_matrix", "GENEMABR")
   }

  all_genes=rownames(gene_pathway_matrix)
  all_pathways=colnames(gene_pathway_matrix)

  module_labels=rep(0,length(all_genes))          #len:20244F
  names(module_labels)=all_genes
  module_common_genes=intersect(all_genes,gene_input)

  if(length(module_common_genes)>1){                     # should set a lower thereshold for num of module common genes, more than 50%
    module_labels[module_common_genes]=1
    if(length(addi_args)==0){
      cvfit=glmnet(gene_pathway_matrix,module_labels,lambda = lambda,alpha =alpha,...)
    }else{
      cvfit=glmnet(gene_pathway_matrix,module_labels,alpha =alpha,...)
    }
    coef=coef(cvfit, s = "lambda.min")
    non0index=coef@i[-1]   #remove intercept
    non0coef=coef@x[-1]
    selected_index=non0index[which(non0coef>0)]
    selected_pathways=all_pathways[selected_index]
    selected_coef=non0coef[which(non0coef>0)]
    names(selected_coef)=selected_pathways

    if(length(selected_pathways)>0){
      fisher_exact_test_results=fisher_exact_test(selected_pathways,module_common_genes,gene_pathway_matrix="default" )
      selected_pathways_fisher_pvalue=fisher_exact_test_results$selected_pathways_fisher_pvalue
      selected_pathways_num_genes=fisher_exact_test_results$selected_pathways_num_genes

      new_order=order(selected_coef,decreasing = TRUE)

      return(list(selected_pathways_names=from_id2name(selected_pathways=names(selected_coef[new_order])),
                  selected_pathways_coef=selected_coef[new_order],
                  selected_pathways_fisher_pvalue=selected_pathways_fisher_pvalue[new_order],
                  selected_pathways_num_genes=selected_pathways_num_genes[new_order]
                  ))
    }else{
      return(NULL)}
  }

}




R/supplementary_functions.R
#' Helper function to load file
#' @param objName Object name to be loaded
#' @param package Package name to be loaded
#' @return The object
#' @importFrom utils data
#' This function is temporarily set up to help the transition from .rds to .RData file
mydata <- function(objName, package) {
    nv <- new.env()
    data(list=c(objName), package=package, envir = nv)
    return(get(objName, envir = nv))
}

#' fisher_exact_test
#'
#' This function allows you to compute two sided fish exact pvalue of gene list for selected  pathways
#' To know more about this method. I recommend you to read the paper (Enrichment or depletion of a GO category within a class of genes: which test?) for more details
#' @param selected_pathways A vecor of pathways to be used for enrichment analysis for genes in \emph{gene_input}.It should have same ID types(E.g. pathway ID, pathway names) as the pathways in \emph{gene_pathway_matrix}.
#' @param gene_input A vecor of genes to be annotated. It should have same ID types(E.g. Ensembl ID, HUGO gene symbol) as the genes in \emph{gene_pathway_matrix}.
#' @param gene_pathway_matrix A binary background matrix whose columns are the pathways/gene sets and
#'whose rows are all the genes from pathways/gene sets . It could be in sparse matrix format ((inherit from class "sparseMatrix" as in package Matrix) to save memory.
#'For gene i and pathway j, the value of matrix(i,j) is 1 is gene i belonging to pathway j otherwise 0.
#'Users could leave it as default value("default") so it will use pre-collected gene_pathway_matrix from GO Ontology and REACTOME databaase.
#'Otherwise, they could use their own customized gene_pathway_matrix
#' @return  A list of two elements:
#' \itemize{
#'   \item selected_pathways_fisher_pvalue - Fisher exact pvalue for selected pathways
#'   \item selected_pathways_num_genes - The number of genes for selected pathways in background
#' }
#' @export
#' @examples
#' fetRes <- fisher_exact_test(selected_pathways=c("GO:0007250","GO:0008625"),
#'   gene_input=c("TRPC4AP","CDC37","TNIP1","IKBKB","NKIRAS2","NFKBIA","TIMM50",
#'      "RELB","TNFAIP3","NFKBIB","HSPA1A","NFKBIE","SPAG9","NFKB2","ERLIN1",
#'      "REL","TNIP2","TUBB6","MAP3K8"),gene_pathway_matrix="default")
fisher_exact_test=function(selected_pathways,gene_input,gene_pathway_matrix="default"){
  if(gene_pathway_matrix=="default"){
    gene_pathway_matrix=mydata("gene_pathway_matrix", package="GENEMABR")
  }
  all_genes=rownames(gene_pathway_matrix)
  module1_common_genes=intersect(all_genes,gene_input)

  selected_pathways_fisher_pvalue=vector()
  selected_pathways_num_genes=vector()
  for(index in seq_len(length(selected_pathways))){
    fisher_pathway=selected_pathways[index]
    #          genes_in_common      genes_in_reference
    #pathway          a                   c
    #no_pathway       b                   d
    a=sum(gene_pathway_matrix[module1_common_genes,fisher_pathway])
    b=length(module1_common_genes)-a
    c=sum(gene_pathway_matrix[,fisher_pathway])-a
    d=length(all_genes)-a-c-b
    contigency_table=matrix(c(a,b,c,d),nrow = 2)
    fisher_result=stats::fisher.test(contigency_table)
    selected_pathways_fisher_pvalue[fisher_pathway]=fisher_result[['p.value']]
    selected_pathways_num_genes[fisher_pathway]=sum(gene_pathway_matrix[,fisher_pathway])
  }
  return(list(selected_pathways_fisher_pvalue=selected_pathways_fisher_pvalue,selected_pathways_num_genes=selected_pathways_num_genes))
}



#' find_root_ids
#'
#' If you use the default pathway databases(GO Ontologyand REACTOME),this function allows you to extract GO sub-roots or REACTOME roots for certain pathways from GO or REACTOME
#' to help you better understanding thier the biological meanings
#' @param selected_pathways A vecor of GO and/or REACTOME pathways IDs.
#' @return  A list of GO sub-root or REACTOME root ids for provided pathways.
#' If a certain pathway has morn than one GO sub-roots or REACTOME roots, they will be seperated by "#".
#' @importFrom igraph ego gorder
#' @export
#' @examples
#' find_root_ids(selected_pathways=c("GO:0005834","R-HSA-111469"))
find_root_ids=function(selected_pathways){
  go_ontology=mydata("human_go_ontology", package="GENEMABR")
  go_sub_roots=mydata("human_go_sub_roots", package="GENEMABR")
  go_sub_roots=unlist(go_sub_roots)
  reactome_ontology=mydata("human_reactome_ontology", package="GENEMABR")
  reactome_roots=mydata("human_reactome_roots", package="GENEMABR")

  find_roots=lapply(selected_pathways, function(x){
    if(grepl("GO",x)){
      parent_nodes=ego(go_ontology,order = gorder(go_ontology),x,mode = "in")
      find_roots=intersect(go_sub_roots,names(unlist(parent_nodes)))
    }else{
      parent_nodes=ego(reactome_ontology,order = gorder(reactome_ontology),x,mode = "in")
      find_roots= intersect(reactome_roots,names(unlist(parent_nodes)))
    }
    if(length(find_roots)>1){
      find_roots=paste0(unlist(find_roots),collapse = "#")
    }
    return(find_roots)
  })
  names(find_roots)=selected_pathways
  return(find_roots)

}


#' from_id2name
#'
#' If you use the default pathway databases(GO Ontologyand REACTOME),this function can help you to get pathways names from pathways IDs.
#' @param selected_pathways A list of GO and/or REACTOME pathways IDs. Each elmment is this list can be a single id or multi-ids seperated "#"
#' @return  A list of GO sub-root or REACTOME root names for provided pathways.
#' @importFrom igraph V as_ids
#' @export
#' @examples
#' from_id2name((selected_pathways=list(c("GO:0032991#GO:0044425#GO:0044464"),"R-HSA-5357801")))
from_id2name=function(selected_pathways){
  go_ontology=mydata("human_go_ontology", package="GENEMABR")
  reactome_ontology=mydata("human_reactome_ontology", package="GENEMABR")

  go_ontology_names=V(go_ontology)$pathway_names
  names(go_ontology_names)=as_ids(V(go_ontology))
  reactome_ontology_names=V(reactome_ontology)$pathway_names
  names(reactome_ontology_names)=as_ids(V(reactome_ontology))

  names=lapply(selected_pathways, function(x){
    if(grepl("#",x)){
      if(grepl("GO",x)){
        y=strsplit(x,split = "#")[[1]]
        unname(go_ontology_names[y])
      }else{
        y=strsplit(x,split = "#")[[1]]
        unname(reactome_ontology_names[y])
      }
    }else{
      if(grepl("GO",x)){
        unname(go_ontology_names[x])
      }else{
        unname(reactome_ontology_names[x])
      }
    }
  })
  names(names)=selected_pathways
  return(names)
}

#' get_steps
#'
#' If you use the default pathway databases(GO Ontologyand REACTOME),this function allows you to extract  the distances from ceatain pathways to  GO roots or REACTOME roots nodes.
#' @param selected_pathways A vecor of GO and/or REACTOME pathways IDs.
#' @return  A list contains distances from pathways to GO root or REACTOME root nodes
#' @importFrom igraph distances
#' @export
#' @examples
#' get_steps(selected_pathways=c("GO:0005834","R-HSA-111469"))
get_steps=function(selected_pathways){
  go_ontology=mydata("human_go_ontology", package="GENEMABR")
  go_roots=mydata("human_go_roots", package = "GENEMABR")
  go_roots=unlist(go_roots)
  reactome_ontology=mydata("human_reactome_ontology", package="GENEMABR")
  reactome_roots=mydata("human_reactome_roots", package="GENEMABR")

  steps=lapply(selected_pathways,function(x){
    if(grepl("GO",x)){
      min(distances(go_ontology,v=go_roots,to=x))
    }else{
      min(distances(reactome_ontology,v=reactome_roots,to=x))
    }
  })
  names(steps)=selected_pathways
  return(steps)
}

